### PR TITLE
fix: remove stale instance entry on recreate and guard stop against missing containers

### LINKED ISSUE #483: bug: stale instance entries accumulate in config when container is recreated
## Description

When a container is removed externally (e.g. `docker stop && docker rm`) and
`microcks start` recreates it, the old instance entry is never removed from config.
A new entry is appended instead, causing duplicate instance records to accumulate.

### Expected behavior

_No response_

### Actual behavior

_No response_

## Steps to Reproduce

1. `./microcks start`
2. `docker stop microcks && docker rm microcks`
3. `./microcks start`
4. `cat ~/.config/microcks/config`  two instance entries with same name but different container IDs

## Screenshots

Image

Image

## Root Cause

`RemoveInstance` matched by `Name` instead of `ContainerID`. Since the recreated
container has a new ID, `UpsertInstance` never finds a match and always appends.

### Microcks version or git rev

_No response_

### Install method (`docker-compose`, `helm chart`, `operator`, `docker-desktop extension`,...)

_No response_

### Additional information

_No response_

### PR DESCRIPTION
**Pull request by Syedowais312**
## Problem

When a container is removed externally (e.g. `docker system prune`) and
`microcks start` recreates it, the old instance entry was never removed from
config  causing duplicate instance records to accumulate on every recreate.

Two root causes:
1. `RemoveInstance` matched by `Name` (not unique) instead of `ContainerID` (unique)
2. The `autoRemove` path in `stop.go` was passing `instance.Name` to `RemoveInstance`,
   silently failing to remove the correct entry
3. `start.go` was clearing `instance.ContainerID = ""` instead of calling
   `RemoveInstance` to actually remove the stale record from config

## Changes

- `cmd/start.go`: call `localConfig.RemoveInstance(instance.ContainerID)` to
  remove the stale record before recreating, instead of just clearing the local variable
- `cmd/stop.go`: pass `instance.ContainerID` to `RemoveInstance` in the `autoRemove` path
- `pkg/config/localconfig.go`: `RemoveInstance` now matches by `ContainerID` instead of `Name`

## Testing

Ran `docker rm <container>` while config shows `status: Running`, then
`microcks start --name <instance>` config shows a single clean entry with
the new container ID, no duplicates.

Realted: https://github.com/microcks/microcks-cli/pull/456

Fix: https://github.com/microcks/microcks-cli/issues/483

### CODE REVIEW

**Comment by Vaishnav88sk** [comment]
Matching by `ContainerID` instead of `Name` is the correct approach. Could you add a test case for the duplicate-entry scenario?

**Comment by Syedowais312** [comment]
> Matching by `ContainerID` instead of `Name` is the correct approach. Could you add a test case for the duplicate-entry scenario?

i have added  appropriate test cases in pkg/config/localconfig_test.go

File: pkg/config/localconfig_test.go
Comment (Vaishnav88sk) [inline pkg/config/localconfig_test.go:1]: Please add the standard Apache license header to this new test file for consistency with the rest of the repository.

**Comment by Vaishnav88sk** [comment]
Thanks for adding tests after the earlier review feedback.

One thing I want to call out: the PR title mentions “guard stop against missing containers”, but I don’t see an actual missing-container guard added in `stop.go` in this diff. So:
- add the missing guard behavior as well.

**Comment by Syedowais312** [comment]
> Thanks for adding tests after the earlier review feedback.
> 
> One thing I want to call out: the PR title mentions “guard stop against missing containers”, but I don’t see an actual missing-container guard added in `stop.go` in this diff. So:


sorry i missed it earlier.
i have added a checks to verify whether the container still exists before trying to stop it.

**Review by Vaishnav88sk — CHANGES_REQUESTED** [review]

File: pkg/config/localconfig_test.go
Comment (Syedowais312) [reply pkg/config/localconfig_test.go:1]: [060992b](https://github.com/microcks/microcks-cli/commit/060992b1eee67a9e36e7b90cc3371e445012e773) added

File: cmd/stop.go [hunk 1]
```diff
@@ -40,6 +40,14 @@ func NewStopCommand(globalClientOpts *connectors.ClientOptions) *cobra.Command {
 			errors.CheckError(err)
 			defer containerClient.CloseClient()
 
+			exists, _ := containerClient.ContainerExists(instance.ContainerID)
+			errors.CheckError(err)
```
Comment (Vaishnav88sk) [inline cmd/stop.go]: There is a bug here: `ContainerExists(...)` returns `(bool, error)`, but the error is being discarded with `_`, and then `errors.CheckError(err)` is checking the previous `err` value instead.

Please capture the returned error here and validate that one.

Comment (Vaishnav88sk) [reply cmd/stop.go]: Also line ~70, 
This should stay as `RemoveInstance(instance.ContainerID)`, not `RemoveInstance(instance.Name)`.

Switching it back to `instance.Name` reintroduces the original issue this PR is fixing, since `RemoveInstance()` was changed to match by `ContainerID`.

**Review by Vaishnav88sk — CHANGES_REQUESTED** [review]
Please resolve that. After that LGTM. 👍🏻

File: cmd/stop.go (hunk 1 above)
Comment (Syedowais312) [reply cmd/stop.go]: I have made the require changes @Vaishnav88sk

**Comment by Caesarsage** [comment]
Thanks @Syedowais312 

The reconcile logic on master resets the stale entry in place (Status/ContainerID cleared, then reused) rather than appending a new record — see the "Reconcile before trusting it" block in cmd/start.go.

Could you re-verify duplicates still accumulate on a current build? If not, this can be closed.

**Comment by Syedowais312** [comment]
> Thanks @Syedowais312
> 
> The reconcile logic on master resets the stale entry in place (Status/ContainerID cleared, then reused) rather than appending a new record — see the "Reconcile before trusting it" block in cmd/start.go.
> 
> Could you re-verify duplicates still accumulate on a current build? If not, this can be closed.

Hey @Caesarsage, I pulled the latest `master` and verified again. The bug still seems to exist.

 The reason is around this line:

  https://github.com/microcks/microcks-cli/blob/0dd8f55ba4aa67b3c6e9918ad29bd5b3d6b9dbc5/cmd/start.go#L102

  Here we clear `instance.Status` and `instance.ContainerID` when the old container no longer exists. Because the status becomes empty, the switch falls through to the  `default` case and creates a new container.

  But the old instance entry is still present in the config with the old container ID. Later, `UpsertInstance` compares using `ContainerID`, so the recreated instance has a new ID and gets appended instead of replacing the old entry.

  In this PR, I removed the stale instance before recreating it. I also changed `RemoveInstance` to remove by `ContainerID`, because that is the exact stale entry we are trying to clean up.

**Comment by Syedowais312** [comment]
Hey, I've rebased the branch onto the latest master. Could you take a look when you get a chance?

File: cmd/stop.go [hunk 2]
```diff
@@ -62,6 +62,16 @@ func NewStopCommand(globalClientOpts *connectors.ClientOptions) *cobra.Command {
...
+			}
+			if !exists {
+				fmt.Printf("Container for instance %s no longer exists\n", instance.Name)
+				fmt.Printf("Run 'microcks start --name %s' to bring the container back\n", instance.Name)
+				return nil
+			}
+
 			if err := containerClient.StopContainer(instance.ContainerID); err != nil {
 				return errors.Wrap(errors.KindEnvironment, fmt.Errorf("failed to stop container: %w", err))
 			}
```
Comment (Vaishnav88sk) [inline cmd/stop.go]: Returning `nil` early here prevents the CLI from crashing (which is a great improvement over the old behavior where `StopContainer` would throw an error), but it completely bypasses the configuration cleanup that happens at the bottom of the file.
**Suggestion**: Instead of returning early, we should just skip the `StopContainer` call and proceed directly to the `// update configs` section so the local state accurately reflects that the container is gone.

```suggestion
			if !exists {
				fmt.Printf("Container for instance %s no longer exists\n", instance.Name)
			} else {
				if err := containerClient.StopContainer(instance.ContainerID); err != nil {
					return errors.Wrap(errors.KindEnvironment, fmt.Errorf("failed to stop container: %w", err))
				}
				fmt.Println("")
				log.Printf("Instance %s stopped successfully", instance.Name)
			}
```

**Review by Vaishnav88sk — CHANGES_REQUESTED** [review]
